### PR TITLE
implementation of additional arithmetic functions

### PR DESCRIPTION
**Pull request by krajsek**
## Description

This Pull Request introduces a couple of arithmetic functions missing in Heat, including the handling of NaN values and signs.

Issue/s resolved: #892 

## Changes proposed:
- nanprod
- nansum
- lcm
- hypot
- nan_to_num
- copysign
- gcd
- Use `True` as the default for `where`

## Type of change

- New feature (non-breaking change which adds functionality)

## Due Diligence
- [ ] All split configurations tested
- [ ] Multiple dtypes tested in relevant functions
- [ ] Documentation updated (if needed)
- [ ] Updated changelog.md under the title "Pending Additions"

#### Does this change modify the behaviour of other functions? If so, which?
no

### CODE REVIEW

File: heat/core/arithmetics.py [hunk 1]
```diff
@@ -577,6 +653,38 @@ def invert(a: DNDarray, out: DNDarray = None) -> DNDarray:
...
+        if required) must match that of the `out` array.
+
+    Examples
+    --------
+    >>> a = ht.array([6, 12, 15])
+    >>> b = ht.array([3, 4, 5])
+    >>> ht.lcm(a,b)
+    DNDarray([ 6, 12, 15], dtype=ht.int64, device=cpu:0, split=None)
+    """
+    return _operations.__binary_op(torch.lcm, a, b, out, where)
```
Comment (mrfh92) [inline heat/core/arithmetics.py:685]: Would it make sense to add a check whether the inputs have integer datatype?

File: heat/core/arithmetics.py [hunk 2]
```diff
@@ -542,6 +586,38 @@ def floordiv(t1: Union[DNDarray, float], t2: Union[DNDarray, float]) -> DNDarray
...
+        if required) must match that of the `out` array.
+
+    Examples
+    --------
+    >>> a = a=ht.array([2.])
+    >>> b = b=ht.array([1.,3.,3.])
+    >>> ht.hypot(a,b)
+    DNDarray([2.2361, 3.6056, 3.6056], dtype=ht.float32, device=cpu:0, split=None)
+    """
+    return _operations.__binary_op(torch.hypot, a, b, out, where)
```
Comment (mrfh92) [inline heat/core/arithmetics.py:618]: Actually, I am not really sure what the typical use-case for this function is, but would it make sense to add a check whether the inputs are positive?

File: heat/core/tests/test_arithmetics.py
Comment (mrfh92) [inline heat/core/tests/test_arithmetics.py:1]: As far as I can see, all the new tests are for non-distributed arrays only. Can we transfer them to the distributed setting, e.g., by using the currently used (global) entries as local entries?

**Review by mrfh92 — COMMENTED** [review]
Looks fine to me; I only have three comments.

File: heat/core/arithmetics.py (hunk 1 above)
Comment (mtar) [reply heat/core/arithmetics.py:685]: The error message from PyTorch is clear for once, so I didn't bother to add it.
> RuntimeError: "lcm_cpu" not implemented for 'Float'

File: heat/core/arithmetics.py (hunk 2 above)
Comment (mtar) [reply heat/core/arithmetics.py:618]: One example is calculating the length of a 2d vector. I will add a sentence with the equivalent formula for better understanding.

File: heat/core/tests/test_arithmetics.py
Comment (mtar) [reply heat/core/tests/test_arithmetics.py:1]: Originally, I wanted to save some time and make the tests even shorter as they don't put anything new to the table. They call binary_op which is already tested extensively in the unit tests for the other arithmetic functions.

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
This is very helpful @mtar , thanks a lot. I only have a few minor comments.

@krajsek is anything missing here?

**Comment by ClaudiaComito** [comment]
@mtar I think the test error is due to #1005

File: heat/core/arithmetics.py [hunk 3]
```diff
@@ -577,6 +653,38 @@ def invert(a: DNDarray, out: DNDarray = None) -> DNDarray:
...
+) -> DNDarray:
+    """
+    Returns the lowest common multiple of |a| and |b|
+
+    Parameters
+    ----------
+    a:   DNDarray
+         The first input array
+    b:   DNDarray
+         the second input tensor
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/arithmetics.py]: I put it in the description.

File: heat/core/arithmetics.py [hunk 4]
```diff
@@ -688,6 +796,135 @@ def mul(t1: Union[DNDarray, float], t2: Union[DNDarray, float]) -> DNDarray:
...
+        the datatype of the output values will be cast if necessary.
+
+    Examples
+    --------
+    >>> x = ht.array([float('nan'), float('inf'), -float('inf')])
+    >>> ht.nan_to_num(x)
+    nan = 0.0
+    posinf = None
+    neginf = None
+    result tensor([ 0.0000e+00,  3.4028e+38, -3.4028e+38])
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/arithmetics.py]: Suppose not

File: heat/core/tests/test_arithmetics.py [hunk 5]
```diff
@@ -192,6 +192,23 @@ def test_bitwise_xor(self):
...
+        self.assertAlmostEqual(ht.mean(ht.copysign(a, -1.0) + result).item(), 0.0)
+        self.assertAlmostEqual(ht.mean(ht.copysign(a, a) - a).item(), 0.0)
+        self.assertAlmostEqual(ht.mean(ht.copysign(b, b) - b).item(), 0.0)
+        self.assertEquals(ht.copysign(a, 1.0).dtype, ht.float32)
+        self.assertEquals(ht.copysign(b, 1.0).dtype, ht.float32)
+        self.assertNotEqual(ht.copysign(a, 1.0).dtype, ht.int64)
+
+        with self.assertRaises(TypeError):
+            ht.copysign(a, "T")
+            ht.copysign(a, 1j)
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/tests/test_arithmetics.py:211]: These tests were wrong.

File: heat/core/arithmetics.py [hunk 6]
```diff
@@ -40,11 +42,16 @@
     "floordiv",
     "floor_divide",
     "fmod",
```
[the remark this answers is not in the document]
Comment (mtar) [reply heat/core/arithmetics.py:44]: :ballot_box_with_check:

**Review by ClaudiaComito — CHANGES_REQUESTED** [review]
Thanks a  lot @mtar I'm fine with the changes, but something's strange in this branch, `ht.gcd` and `ht.convolve` aren't available (maybe others that I haven't noticed). All is up to date locally, can you reproduce the problem?

| NumPy 1.25  | Heat |
|---|---|
| [Mathematical Functions](https://numpy.org/doc/stable/reference/routines.math.html)|  multi-node, multi-GPU |
| sin | ✅ |
| cos | ✅ |
| tan | ✅ |
| arcsin | ✅ |
| arccos | ✅ |
| arctan | ✅ |
| hypot | ✅ |
| arctan2 | ✅ |
| degrees | ✅ |
| radians | ✅ |
| unwrap | ❌ |
| deg2rad | ✅ |
| rad2deg | ✅ |
| sinh | ✅ |
| cosh | ✅ |
| tanh | ✅ |
| arcsinh | ✅ |
| arccosh | ✅ |
| arctanh | ✅ |
| round | ✅ |
| around | ❌ |
| rint | ❌ |
| fix | ❌ |
| floor | ✅ |
| ceil | ✅ |
| trunc | ✅ |
| prod | ✅ |
| sum | ✅ |
| nanprod | ✅ |
| nansum | ✅ |
| cumprod | ✅ |
| cumsum | ✅ |
| nancumprod | ❌ |
| nancumsum | ❌ |
| diff | ✅ |
| ediff1d | ❌ |
| gradient | ❌ |
| cross | ❌ |
| trapz | ❌ |
| exp | ✅ |
| expm1 | ✅ |
| exp2 | ✅ |
| log | ✅ |
| log10 | ✅ |
| log2 | ✅ |
| log1p | ✅ |
| logaddexp | ✅ |
| logaddexp2 | ✅ |
| i0 | ❌ |
| sinc | ❌ |
| signbit | ✅ |
| copysign | ✅ |
| frexp | ❌ |
| ldexp | ❌ |
| nextafter | ❌ |
| spacing | ❌ |
| lcm | ✅ |
| gcd | ❌ |
| add | ✅ |
| reciprocal | ❌ |
| positive | ✅ |
| negative | ✅ |
| multiply | ✅ |
| divide | ✅ |
| power | ✅ |
| subtract | ✅ |
| true_divide | ❌ |
| floor_divide | ✅ |
| float_power | ❌ |
| fmod | ✅ |
| mod | ✅ |
| modf | ✅ |
| remainder | ✅ |
| divmod | ❌ |
| angle | ✅ |
| real | ✅ |
| imag | ✅ |
| conj | ✅ |
| conjugate | ✅ |
| maximum | ✅ |
| max | ✅ |
| amax | ❌ |
| fmax | ❌ |
| nanmax | ❌ |
| minimum | ✅ |
| min | ✅ |
| amin | ❌ |
| fmin | ❌ |
| nanmin | ❌ |
| convolve | ❌ |
| clip | ✅ |
| sqrt | ✅ |
| cbrt | ❌ |
| square | ✅ |
| absolute | ✅ |
| fabs | ✅ |
| sign | ✅ |
| heaviside | ❌ |
| nan_to_num | ✅ |
| real_if_close | ❌ |
| interp | ❌ |

Code (thanks ChatGPT):
```python
import heat

# List of numpy functions
numpy_functions = [
    "sin", "cos", "tan", "arcsin", "arccos", "arctan", "hypot", "arctan2",
    "degrees", "radians", "unwrap", "deg2rad", "rad2deg", "sinh", "cosh",
    "tanh", "arcsinh", "arccosh", "arctanh", "round", "around", "rint", "fix",
    "floor", "ceil", "trunc", "prod", "sum", "nanprod", "nansum", "cumprod",
    "cumsum", "nancumprod", "nancumsum", "diff", "ediff1d", "gradient", "cross",
    "trapz", "exp", "expm1", "exp2", "log", "log10", "log2", "log1p", "logaddexp",
    "logaddexp2", "i0", "sinc", "signbit", "copysign", "frexp", "ldexp", "nextafter",
    "spacing", "lcm", "gcd", "add", "reciprocal", "positive", "negative",
    "multiply", "divide", "power", "subtract", "true_divide", "floor_divide",
    "float_power", "fmod", "mod", "modf", "remainder", "divmod", "angle", "real",
    "imag", "conj", "conjugate", "maximum", "max", "amax", "fmax", "nanmax",
    "minimum", "min", "amin", "fmin", "nanmin", "convolve", "clip", "sqrt", "cbrt",
    "square", "absolute", "fabs", "sign", "heaviside", "nan_to_num", "real_if_close",
    "interp"
]
print("numpy functions: ", len(numpy_functions))
# Initialize a list to store the rows of the Markdown table
table_rows = []

# Check if functions exist in the heat library and create table rows
for func_name in numpy_functions:
    if hasattr(heat, func_name):
        support_status = "✅"  # Green checkmark for supported functions
    else:
        support_status = "❌"  # Red cross for unsupported functions

    table_row = f"| {func_name} | {support_status} |"
    table_rows.append(table_row)

# Create the Markdown table header
table_header = "| NumPy 1.25 Mathematical Functions | Heat |\n|---|---|\n"

# Combine the header and table rows
markdown_table = table_header + "\n".join(table_rows)

# Print the Markdown table
print(markdown_table)
```

**Comment by mtar** [comment]
Both functions have a green check mark when I run it on my local machine.

**Review by ClaudiaComito — APPROVED** [review]
Perfect, thanks @mtar !

**Pull request merged**